### PR TITLE
Premint - upgradeable premint, new deployment scripts, and premint proxy deployment

### PR DESCRIPTION
**Pull request by oveddan**
New deployment scripts and deployments for premint.
* Changed `DeployPreminter.s.sol` to `script/DeployNewPreminterAndFactoryProxy.s.sol` and extracted common deployment functions `deployNew1155AndFactoryProxy` and `deployNewPreminterProxy` to helper methods in the base deployment contract, allowing for more composability/reusability in deployment scripts.  Updated the base `Deploy.s.sol` to use these helper methods.
* Added an `UpgradePreminter.s.sol` script that deploys a new preminter version and updates the proxy

### CODE REVIEW

**Comment by oveddan** [comment]
Current dependencies on/for this PR:
* main
  * **PR #184** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/184" target="_blank">Graphite</a> 
    * **PR #161** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/161" target="_blank">Graphite</a> 
      * **PR #189** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/189" target="_blank">Graphite</a>  👈
        * **PR #190** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/190" target="_blank">Graphite</a> 
      * **PR #188** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/188" target="_blank">Graphite</a> 
      * **PR #187** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/187" target="_blank">Graphite</a> 
      * **PR #179** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/179" target="_blank">Graphite</a> 
        * **PR #171** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/171" target="_blank">Graphite</a> 
          * **PR #183** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/183" target="_blank">Graphite</a> 
          * **PR #178** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/178" target="_blank">Graphite</a> 
          * **PR #175** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/175" target="_blank">Graphite</a> 
          * **PR #173** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/173" target="_blank">Graphite</a> 
          * **PR #172** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/172" target="_blank">Graphite</a> 
      * **PR #170** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/170" target="_blank">Graphite</a> 
      * **PR #168** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/168" target="_blank">Graphite</a> 
        * **PR #169** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/169" target="_blank">Graphite</a> 
      * **PR #166** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/166" target="_blank">Graphite</a> 
      * **PR #164** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/164" target="_blank">Graphite</a> 
      * **PR #163** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/163" target="_blank">Graphite</a> 
      * **PR #162** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/162" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/189?utm_source=stack-comment).

**Comment by oveddan** [comment]
ToDo: 
* [ ] how do you upgrade both erc1155 and preminter?

File: src/nft/ZoraCreator1155Impl.sol [hunk 1]
```diff
@@ -722,4 +758,43 @@ contract ZoraCreator1155Impl is
             revert();
         }
     }
+
+    /* start eip712 functionality */
+    mapping(uint32 => uint256) public delegatedTokenId;
```
Comment (iainnash) [inline src/nft/ZoraCreator1155Impl.sol:763]: This should be in a storage contract.

Comment (oveddan) [reply src/nft/ZoraCreator1155Impl.sol:763]: done in: #188

File: .storage-layout [hunk 2]
```diff
@@ -33,6 +33,8 @@
 | permissions              | mapping(uint256 => mapping(address => uint256))                          | 510  | 0      | 32    | src/nft/ZoraCreator1155Impl.sol:ZoraCreator1155Impl |
 | __gap                    | uint256[50]                                                              | 511  | 0      | 1600  | src/nft/ZoraCreator1155Impl.sol:ZoraCreator1155Impl |
 | createReferrals          | mapping(uint256 => address)                                              | 561  | 0      | 32    | src/nft/ZoraCreator1155Impl.sol:ZoraCreator1155Impl |
+| firstMinters             | mapping(uint256 => address)                                              | 562  | 0      | 32    | src/nft/ZoraCreator1155Impl.sol:ZoraCreator1155Impl |
```
[the remark this answers is not in the document]
Comment (oveddan) [reply .storage-layout:36]: this was merged into premint #162

**Comment by oveddan** [comment]
this diff is really big let me fix

**Comment by oveddan** [comment]
@iainnash PR updated to just be this diff.  it needed to be restacked

File: .storage-layout (hunk 2 above)
[the remark this answers is not in the document]
Comment (oveddan) [reply .storage-layout]: this diff was accidentally really big for this PR when you reviewed it; just updated the diff to just be for this PR.  if you want to see the full premint massive PR check out #161

**Pull request merged**